lib/async_hooks.js

```diff
@@ -204,11 +206,111 @@ class AsyncResource {
   }
 }
 
+const locals = [];
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    for (let i = 0; i < locals.length; ++i) {
+      locals[i]._propagate(resource, executionAsyncResource());
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
+    asyncContextHook.enable();
+    if (!locals.includes(this)) {
+      locals.push(this);
```

Comment:
When are elements removed from `locals`? It seems this is going to leak memory.

Reply:
So locals is behaving like a `Set`, with the current model, it is not possible to remove something from it as there is no way to disable an instance. I will add `enable()/disable()` methods that will handle this.

Reply:
Done in https://github.com/nodejs/node/pull/26540/commits/c996e7ee247e4636b7fba9b73f2f1105a330b9bd